forge/src/executor/inspector/debugger.rs

```diff
@@ -0,0 +1,202 @@
+use crate::{
+    debugger::{DebugArena, DebugNode, DebugStep, Instruction},
+    executor::CHEATCODE_ADDRESS,
+};
+use bytes::Bytes;
+use ethers::{
+    types::Address,
+    utils::{get_contract_address, get_create2_address},
+};
+use revm::{
+    CallInputs, CreateInputs, CreateScheme, Database, EVMData, Gas, Inspector, Interpreter, Memory,
+    OpCode, Return,
+};
+use std::collections::BTreeMap;
+
+/// An inspector that collects debug nodes on every step of the interpreter.
+#[derive(Default, Debug)]
+pub struct Debugger {
+    /// The arena of [DebugNode]s
+    pub arena: DebugArena,
+    /// The ID of the current [DebugNode].
+    pub head: usize,
+    /// A mapping of program counters to instruction counters.
+    ///
+    /// The program counter keeps track of where we are in the contract bytecode as a whole,
+    /// including push bytes, while the instruction counter ignores push bytes.
+    ///
+    /// The instruction counter is used in Solidity source maps.
+    pub ic_map: BTreeMap<Address, BTreeMap<usize, usize>>,
+}
+
+impl Debugger {
+    pub fn new() -> Self {
+        Default::default()
+    }
+
+    /// Builds the instruction counter map for the given bytecode.
+    // TODO: Some of the same logic is performed in REVM, but then later discarded. We should
+    // investigate if we can reuse it
+    pub fn build_ic_map(&mut self, address: &Address, code: &Bytes) {
+        let mut ic_map: BTreeMap<usize, usize> = BTreeMap::new();
+
+        let mut i = 0;
+        let mut cumulative_push_size = 0;
+        while i < code.len() {
+            let op = code[i];
+            ic_map.insert(i, i - cumulative_push_size);
+            match OpCode::is_push(op) {
+                Some(push_size) => {
+                    // Skip the push bytes
+                    i += push_size as usize;
+                    cumulative_push_size += push_size as usize;
+                }
+                None => (),
+            }
+            i += 1;
+        }
+
+        self.ic_map.insert(*address, ic_map);
+    }
+
+    /// Enters a new execution context.
+    pub fn enter(&mut self, depth: usize, address: Address, creation: bool) {
+        self.head =
+            self.arena.push_node(DebugNode { depth, address, creation, ..Default::default() });
+    }
+
+    /// Exits the current execution context, replacing it with the previous one.
+    pub fn exit(&mut self) {
+        if let Some(parent_id) = self.arena.arena[self.head].parent {
+            let DebugNode { depth, address, creation, .. } = self.arena.arena[parent_id];
+            self.head =
+                self.arena.push_node(DebugNode { depth, address, creation, ..Default::default() });
+        }
+    }
```

Comment:
Called by `call_end` and `create_end` since we go back to the previous execution context, which in this case is the parent of our current head node